File: Medium/JavaScript/486. Predict the Winner.js
var PredictTheWinner = function(nums) {
    let n = nums.length;
    let diff = new Array(n).fill(0);

    for(let i = n - 1; 0 <= i; i--){
        diff[i] = nums[i];
        for(let j = i + 1; j < n; j++){
            diff[j] = Math.max(nums[i] - diff[j], nums[j] - diff[j - 1]);
        }
    }

    return 0 <= diff[n - 1];
};

File: Medium/Java/486. Predict the Winner.java
class Solution {
    public boolean PredictTheWinner(int[] nums) {
        int n = nums.length;
        int[] diff = new int[n];

        for(int i = n - 1; 0 <= i; i--){
            diff[i] = nums[i];
            for(int j = i + 1; j < n; j++){
                diff[j] = Math.max(nums[i] - diff[j], nums[j] - diff[j - 1]);
            }
        }

        return 0 <= diff[n - 1];
    }
}
